### PR TITLE
fix #34159, crash in deserialize caused by recent layout changes

### LINKED ISSUE #34159: seg fault `using Gadfly` on 0-day master
works fine on 1.3, and a week ago it worked fine on a 0-day master, but now it's broken.  this is on MacOS.  a transcript follows:

```
$ src/julia/usr/bin/julia 
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.4.0-DEV.648 (2019-12-20)
 _/ |\__'_|_|_|\__'_|  |  Commit 608567f856 (0 days old master)
|__/                   |

(@v1.4) pkg> st
Status `~/.julia/environments/v1.4/Project.toml`
  (empty environment)

(@v1.4) pkg> dev Gadfly Compose
Path `/Users/arthurb/.julia/dev/Gadfly` exists and looks like the correct package. Using existing path.
Path `/Users/arthurb/.julia/dev/Compose` exists and looks like the correct package. Using existing path.
  Updating `~/.julia/environments/v1.4/Project.toml`
  [a81c6b42] + Compose v0.7.4 [`~/.julia/dev/Compose`]
  [c91e804a] + Gadfly v1.1.0 [`~/.julia/dev/Gadfly`]
  Updating `~/.julia/environments/v1.4/Manifest.toml`
  [621f4979] + AbstractFFTs v0.5.0
  [7d9fca2a] + Arpack v0.4.0
  [68821587] + Arpack_jll v3.5.0+2
  [4fba245c] + ArrayInterface v2.1.0
  [13072b0f] + AxisAlgorithms v1.0.0
  [b99e7846] + BinaryProvider v0.5.8
  [49dc2e85] + Calculus v0.5.1
  [324d7699] + CategoricalArrays v0.7.4
  [3da002f7] + ColorTypes v0.8.1
  [5ae59095] + Colors v0.9.6
  [bbf7d656] + CommonSubexpressions v0.2.0
  [34da2185] + Compat v2.2.0
  [a81c6b42] + Compose v0.7.4 [`~/.julia/dev/Compose`]
  [d38c429a] + Contour v0.5.1
  [7ad07ef1] + CoupledFields v0.1.0
  [9a962f9c] + DataAPI v1.1.0
  [864edb3b] + DataStructures v0.17.6
  [01453d9d] + DiffEqDiffTools v1.6.0
  [163ba53b] + DiffResults v1.0.1
  [b552c78f] + DiffRules v1.0.0
  [b4f34e82] + Distances v0.8.2
  [31c24e10] + Distributions v0.21.11
  [ffbed154] + DocStringExtensions v0.8.1
  [7a1cc6ca] + FFTW v1.2.0
  [f5851436] + FFTW_jll v3.3.9+3
  [1a297f60] + FillArrays v0.8.2
  [53c48c17] + FixedPointNumbers v0.6.1
  [f6369f11] + ForwardDiff v0.10.8
  [c91e804a] + Gadfly v1.1.0 [`~/.julia/dev/Gadfly`]
  [a1b4810d] + Hexagons v0.2.0
  [9b13fd28] + IndirectArrays v0.5.1
  [1d5cc7b8] + IntelOpenMP_jll v2018.0.3+0
  [a98d9a8b] + Interpolations v0.12.5
  [c8e1da08] + IterTools v1.3.0
  [682c06a0] + JSON v0.21.0
  [e5e0dc1b] + Juno v0.7.2
  [5ab0869b] + KernelDensity v0.5.1
  [d3d80556] + LineSearches v7.0.1
  [4345ca2d] + Loess v0.5.0
  [856f044c] + MKL_jll v2019.0.117+0
  [1914dd2f] + MacroTools v0.5.3
  [442fdcdd] + Measures v0.3.1
  [e89f7d12] + Media v0.5.0
  [e1d29d7a] + Missings v0.4.3
  [d41bc354] + NLSolversBase v7.5.0
  [77ba4419] + NaNMath v0.3.3
  [6fe1bfb0] + OffsetArrays v0.11.4
  [4536629a] + OpenBLAS_jll v0.3.7+1
  [efe28fd5] + OpenSpecFun_jll v0.5.3+1
  [429524aa] + Optim v0.19.7
  [bac558e1] + OrderedCollections v1.1.0
  [90014a1f] + PDMats v0.9.10
  [d96e819e] + Parameters v0.12.0
  [69de0a69] + Parsers v0.3.10
  [85a6dd25] + PositiveFactorizations v0.2.3
  [1fd47b50] + QuadGK v2.3.1
  [c84ed2f1] + Ratios v0.3.1
  [189a3867] + Reexport v0.2.0
  [ae029012] + Requires v1.0.0
  [79098fc4] + Rmath v0.6.0
  [992d4aef] + Showoff v0.3.1
  [a2af1166] + SortingAlgorithms v0.3.1
  [276daf66] + SpecialFunctions v0.9.0
  [90137ffa] + StaticArrays v0.12.1
  [2913bbd2] + StatsBase v0.32.0
  [4c63d2b9] + StatsFuns v0.9.3
  [efce3f68] + WoodburyMatrices v0.4.1
  [2a0f44e3] + Base64 
  [ade2ca70] + Dates 
  [8bb1440f] + DelimitedFiles 
  [8ba89e20] + Distributed 
  [9fa8497b] + Future 
  [b77e0a4c] + InteractiveUtils 
  [76f85450] + LibGit2 
  [8f399da3] + Libdl 
  [37e2e46d] + LinearAlgebra 
  [56ddb016] + Logging 
  [d6f4376e] + Markdown 
  [a63ad114] + Mmap 
  [44cfe95a] + Pkg 
  [de0858da] + Printf 
  [9abbd945] + Profile 
  [3fa0cd96] + REPL 
  [9a3f8284] + Random 
  [ea8e919c] + SHA 
  [9e88b42a] + Serialization 
  [1a1011a3] + SharedArrays 
  [6462fe0b] + Sockets 
  [2f01184e] + SparseArrays 
  [10745b16] + Statistics 
  [4607b0f0] + SuiteSparse 
  [8dfed614] + Test 
  [cf7118a7] + UUIDs 
  [4ec0a83e] + Unicode 

julia> using Gadfly
[ Info: Precompiling Gadfly [c91e804a-d5a3-530f-b6f0-dfbca275c004]

signal (11): Segmentation fault: 11
in expression starting at REPL[3]:1
inst_ftypes at /Users/arthurb/src/julia/src/jltypes.c:1378
jl_compute_fieldtypes at /Users/arthurb/src/julia/src/jltypes.c:1610
jl_field_type at /Users/arthurb/src/julia/src/./julia.h:960 [inlined]
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1950 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_code_instance at /Users/arthurb/src/julia/src/dump.c:1809 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2155
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1788 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_method at /Users/arthurb/src/julia/src/dump.c:1716 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2151
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2008 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1490 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_svec at /Users/arthurb/src/julia/src/dump.c:1517 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2080
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1494 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1767 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_value_array at /Users/arthurb/src/julia/src/dump.c:1579
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1785 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_value_array at /Users/arthurb/src/julia/src/dump.c:1579
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1785 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_value_array at /Users/arthurb/src/julia/src/dump.c:1579
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1785 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_method at /Users/arthurb/src/julia/src/dump.c:1716 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2151
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2008 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1490 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_svec at /Users/arthurb/src/julia/src/dump.c:1517 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2080
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1494 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2008 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1490 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_svec at /Users/arthurb/src/julia/src/dump.c:1517 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2080
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1494 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1767 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_value_array at /Users/arthurb/src/julia/src/dump.c:1579
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1785 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_value_array at /Users/arthurb/src/julia/src/dump.c:1579
jl_deserialize_value_method_instance at /Users/arthurb/src/julia/src/dump.c:1785 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2153
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_method at /Users/arthurb/src/julia/src/dump.c:1716 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2151
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2008 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1490 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_module at /Users/arthurb/src/julia/src/dump.c:1852 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2157
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:1999 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1490 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_svec at /Users/arthurb/src/julia/src/dump.c:1517 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2080
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1498 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_svec at /Users/arthurb/src/julia/src/dump.c:1517 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2080
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1494 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_svec at /Users/arthurb/src/julia/src/dump.c:1517 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2080
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1494 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_struct at /Users/arthurb/src/julia/src/dump.c:1938 [inlined]
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2036 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_value_any at /Users/arthurb/src/julia/src/dump.c:2008 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2241
jl_deserialize_datatype at /Users/arthurb/src/julia/src/dump.c:1490 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2238
jl_deserialize_value_singleton at /Users/arthurb/src/julia/src/dump.c:1922 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2195
jl_deserialize_value_module at /Users/arthurb/src/julia/src/dump.c:1852 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2157
jl_deserialize_value_module at /Users/arthurb/src/julia/src/dump.c:1844 [inlined]
jl_deserialize_value at /Users/arthurb/src/julia/src/dump.c:2157
jl_deserialize_value_array at /Users/arthurb/src/julia/src/dump.c:1579
_jl_restore_incremental at /Users/arthurb/src/julia/src/dump.c:3242
jl_restore_incremental at /Users/arthurb/src/julia/src/dump.c:3301
_include_from_serialized at ./loading.jl:681
_require_from_serialized at ./loading.jl:748
_require at ./loading.jl:1039
require at ./loading.jl:927
require at ./loading.jl:922
jl_apply at /Users/arthurb/src/julia/src/./julia.h:1692 [inlined]
call_require at /Users/arthurb/src/julia/src/toplevel.c:399 [inlined]
eval_import_path at /Users/arthurb/src/julia/src/toplevel.c:436
jl_toplevel_eval_flex at /Users/arthurb/src/julia/src/toplevel.c:656
jl_toplevel_eval_flex at /Users/arthurb/src/julia/src/toplevel.c:764
jl_toplevel_eval at /Users/arthurb/src/julia/src/toplevel.c:823 [inlined]
jl_toplevel_eval_in at /Users/arthurb/src/julia/src/toplevel.c:843
eval at ./boot.jl:331
eval_user_input at /Users/arthurb/src/julia/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86
macro expansion at /Users/arthurb/src/julia/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:118 [inlined]
#26 at ./task.jl:358
jl_apply at /Users/arthurb/src/julia/src/./julia.h:1692 [inlined]
start_task at /Users/arthurb/src/julia/src/task.c:687
Allocations: 25131669 (Pool: 25126255; Big: 5414); GC: 23
Segmentation fault: 11

$ sw_vers -productVersion 
10.14.5
```

### PR DESCRIPTION
**Pull request by JeffBezanson**
The field types are not necessarily available at this point in deserialization. I also like this implementation a bit better since it always handles null-ing C pointers on the serialize side.

### CODE REVIEW

File: src/dump.c [hunk 1]
```diff
@@ -1082,25 +1082,36 @@ static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_li
...
+            size_t offs = jl_field_offset(t, i);
+            char *ptr = data + offs;
+            if (t->mutabl && jl_is_cpointer_type(jl_field_type(t, i)) && *(void**)ptr != (void*)-1) {
+                if (ptr > last)
+                    ios_write(s->s, last, ptr - last);
+                char *n = NULL;
+                ios_write(s->s, (char*)&n, sizeof(n));
+                last = ptr + sizeof(n);
+            }
+            else if (j < np && ptr == (char*)&((jl_value_t**)data)[jl_ptr_offset(t, j)]) {
```
Comment (vtjnash) [inline src/dump.c]: I think this needs `while ptr > jl_ptr_offset(t, j) { /* write out previous field */ }`

**Comment by vtjnash** [comment]
Agreed. This is a better approach. I was alternatively also thinking about copying the value into temporary space (`copy = memcpy(malloc(sz), src, sz)`), then iterating fields to clear the pointers inside that copy, then serializing out the copied memory.

File: src/dump.c [hunk 2]
```diff
@@ -1082,25 +1082,45 @@ static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_li
...
+            char *ptr = data + (i < nf ? jl_field_offset(t, i) : jl_datatype_size(t));
+            if (j < np) {
+                char *prevptr = (char*)&((jl_value_t**)data)[jl_ptr_offset(t, j)];
+                while (ptr > prevptr) {
+                    // previous field contained pointers; write them and their interleaved data
+                    if (prevptr > last)
+                        ios_write(s->s, last, prevptr - last);
+                    jl_value_t *e = *(jl_value_t**)prevptr;
+                    JL_GC_PROMISE_ROOTED(e);
+                    if (t->mutabl && e && jl_is_cpointer(e) && jl_unbox_voidpointer(e) != (void*)-1 && jl_unbox_voidpointer(e) != NULL)
```
Comment (vtjnash) [inline src/dump.c]: You probably want `jl_field_type(t, i)->mutabl` here (but that's always false)
```suggestion
                    if (0)
```

Comment (JeffBezanson) [reply src/dump.c]: Oh right, we need some kind of check for whether this is a direct field of the object. Maybe checking that `jl_field_type(t, i-1)` is concrete and inlinealloc?

Comment (vtjnash) [reply src/dump.c]: Ah, right. We probably want `t->mutable && jl_field_isptr(t, i - 1) && jl_is_cpointer(e) &&` then?

**Pull request merged**